soufi/finders/photon.py

```diff
@@ -1,157 +1,69 @@
-        content = self._get_url(url)
-        tree = html.fromstring(content)
-        # Ideally all the SRPM trees would have the exact same packages in
-        # them, but their `aarch64` trees seem to be a little light.  Prefer
-        # x86_64 to be safe
-        dirs = tree.xpath(
+    def _get_repos(self, xpath):
+        dirs = []
+        for release_dir in self._get_dirs():
+            url = f"{PHOTON_PACKAGES}/{release_dir}"
+            try:
+                content = self._get_url(url)
+            except soufi.exceptions.DownloadError:
+                continue
+            tree = html.fromstring(content)
+            # Ideally all the SRPM trees would have the exact same
+            # packages in them, but their `aarch64` trees seem to be a
+            # little light.  Prefer x86_64 to be safe
+            dirs += [url + dir for dir in tree.xpath(xpath)]
+        return dirs
+
+    def _get_source_repos(self):
+        return self._get_repos(
             "//a[text()[contains(.,'srpms')][contains(.,'x86_64')]]/text()"
         )
 
-        for dir_ in dirs:
-            try:
-                _url = f"{url}{dir_}"
-                yield _url, self._get_url(_url)
-            except exceptions.DownloadError:
-                pass
-
-    def _get_path(self, content):
-        """Given a dir, see if it contains the source HREF."""
-        # Grab source directory listing at dir and look for
-        # <name>-<version>.src.rpm.
-        tree = html.fromstring(content)
-        href = f"{self.name}-{self.version}.src.rpm"
-        ref = tree.xpath('//a[@href=$href]', href=href)
-        if ref:
-            return href
-        return None
-
-    def _repo_lookup(self, base_url):
-        content = self._get_url(base_url)
-        tree = html.fromstring(content)
-        os_dirs = tree.xpath(
-            "//a[contains(.,'updates') "
-            "or contains(.,'release') "
-            "or contains(.,'extras')]/text()"
+    def _get_binary_repos(self):
+        return self._get_repos(
+            "//a[text()[not(contains(.,'srpms'))][contains(.,'x86_64')]]/text()"  # noqa: E501
         )
-        for os_dir in os_dirs:
-            bin_url = f"{base_url}{os_dir}"
-            repo = self._get_repo(bin_url)
-            if repo is None:
-                break
-            for package in repo.findall(self.name):
-                if package.evr == self.version:
-                    break
-            else:
-                continue
-            return package.sourcerpm
-        return None
 
-    # Cache repo downloads as they are slow and network-bound.
-    @classmethod
-    @functools.lru_cache(maxsize=128)
-    def _get_repo(cls, url):
-        try:
-            return repomd.load(url)
-        except urllib.error.HTTPError:
-            return None
-
-
-class PhotonDiscoveredSource(finder.DiscoveredSource):
-    """A discovered Photon source package."""
+    def _walk_source_repos(self, name, version=None):
```

Comment:
Looks like the walk* funcs need to be public too.

Reply:
Not really — they're definitely implementation details of the `YumFinder` and aren't supposed to be called or cared about by subclasses.  If you're overriding them, something has gone horribly wrong (e.g., Photon)

Reply:
If something has gone horribly wrong then the design is wrong. Is there any aspect you can factor out and allow that to be overridden?